plugins/remark-lint-no-dead-urls/src/index.js

```diff
@@ -0,0 +1,124 @@
+import {fetch} from './lib.js';
+
+// Forked from: https://github.com/davidtheclark/remark-lint-no-dead-urls
+
+const linkCache = new Map();
+
+const HTTP = {
+  OK: 200,
+  NOT_FOUND: 404,
+  TOO_MANY_REQUESTS: 429,
+};
+
+const uri = {
+  isLocalhost: url => /^(https?:\/\/)(localhost|127\.0\.0\.1)(:\d+)?/.test(url),
+  isExternal: url => /(https?:\/\/)/.test(url),
+  isPath: url => /^\/.*/.test(url),
+};
+
+async function cacheFetch(urlOrPath, method, options) {
+  if (linkCache.has(urlOrPath)) {
+    return [urlOrPath, linkCache.get(urlOrPath)];
+  }
+
+  const {baseUrl, ...other} = options;
+  const url = new URL(urlOrPath, baseUrl).toString();
+
+  const code = await fetch(url, method, other);
+
+  linkCache.set(urlOrPath, code);
+  return [urlOrPath, code];
+}
+
+async function naiveLinkCheck(urls, options) {
+  return Promise.allSettled(
+    urls.map(async url => {
+      try {
+        return await cacheFetch(url, 'HEAD', options);
+      } catch (e) {
+        try {
+          // Fallback, some endpoints don't support HEAD requests
+          return await cacheFetch(url, 'GET', options);
+        } catch (e) {
+          if (e.code === 'ERR_GOT_REQUEST_ERROR') {
+            throw e;
+          }
+          const code = e.statusCode ?? e?.response?.statusCode ?? e.code;
+          linkCache.set(url, code);
+          return [url, code];
+        }
+      }
+    })
+  );
+}
+
+async function noDeadUrls(ast, file, options = {}) {
+  const urlToNodes = new Map();
+
+  const {skipUrlPatterns, ...clientOptions} = options;
+
+  // Grab all possible urls from the markdown
+  visit(ast, ['link', 'image', 'definition'], node => {
+    const {url} = node;
+    if (
+      !url ||
+      uri.isLocalhost(url) ||
+      skipUrlPatterns?.some(skipPattern => new RegExp(skipPattern).test(url))
+    ) {
+      return;
+    }
+
+    // It only makes sense to consider paths when we know the base url. This is useful for images, or cross
+    // references. There might be false positives when adding new pages that aren't already live.
+    const isGoodRelativePath = clientOptions.baseUrl && uri.isPath(url);
+    const isExternalURL = uri.isExternal(url);
+    if (!isExternalURL && !isGoodRelativePath) {
+      return;
+    }
+
+    if (!urlToNodes.has(url)) {
+      urlToNodes.set(url, []);
+    }
+
+    urlToNodes.get(url).push(node);
+  });
+
+  const results = await naiveLinkCheck([...urlToNodes.keys()], clientOptions);
+
+  for (const {value, ...other} of results) {
+    const [url, statusCode] = value;
+    const nodes = urlToNodes.get(url) ?? [];
+
+    if (statusCode === HTTP.OK) {
+      continue;
+    }
+
+    for (const node of nodes) {
+      switch (statusCode) {
+        case 'ENOTFOUND':
+          file.message(`Link to ${url} is broken, domain not found`, node);
+          break;
+        case HTTP.TOO_MANY_REQUESTS:
+          file.message(`Link to ${url} is being rate limited`, node);
+          break;
```

Comment:
Can we retry on this based on backoff? I ran into this before when [scraping from GitHub](https://github.com/rn-versions/rn-versions.github.io/blob/main/scripts/helper/createAxiosInstance.ts#L15), and it should be similar to do something with `got`'s retry. I think.

Reply:
I think we get this out of the box with `got`.  At the moment it will:
- [retry 2 times](https://github.com/sindresorhus/got/blob/0ca0b7f7134f41b45a51370154041cc97c28ca60/source/core/options.ts#L727) with the default configuration, and
- it respects [retry-after](https://github.com/sindresorhus/got/blob/0ca0b7f7134f41b45a51370154041cc97c28ca60/test/retry.ts#L208-L333) without needing additional configuration.

Those seem sane to me.  Thoughts @NickGerleman ?

Reply:
Ah, sounds good, if it already works out of the box.